### DISCUSSION TITLE
Is it OK to create and use SQLite triggers?

### DISCUSSION THREAD

**Discussion by patronovski** [pocketbase/pocketbase#6178] [Q&A] [answered]
Can i create and use SQL triggers or will this interfere with pocketbase?
This question also applies for tables/views, are we safe if we just use plain SQL to create our tables/views?

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
> Can i create and use SQL triggers or will this interfere with pocketbase?

You can but I don't recommend it. You'll have more control and more advanced options if you use the [app event hooks](https://pocketbase.io/docs/go-event-hooks/).
Regarding interference - it depends on the trigger but note that currently the Create API rule executes after a dry submit which may cause unintended side-effects for your trigger action (with the upcoming v0.24 this no longer should be the case, see https://github.com/pocketbase/pocketbase/discussions/6073). In addition to that, manipulating the database with raw SQL queries have the risk of leaving your DB in invalid or even broken state because many app level validators and data integrity checks will not be executed (including cascade deletes/unsets).

> are we safe if we just use plain SQL to create our tables/views?

Again, you can do that but these plain SQL created tables will not appear in the Dashboard or have a Web API access for them because they will not have a corresponding `_collections` table entry.

There are no details about your use case to understand the bigger picture of what you are trying to do and why, but if you mostly prefer and plan to execute plain SQL statements, then PocketBase may NOT be the best choice and you probably would be better with some other solution (e.g. Supabase) or simply use SQLite directly.

**Reply by patronovski** [reply]
Thanks, this is really helpful information. The main reason was because i wanted the transactional behavior of triggers. We can achieve this too with app event hooks, but you need to do a workaround, because its not supported out-of-the-box.

**Reply by ganigeorgiev** [reply]
If you imply https://github.com/pocketbase/pocketbase/discussions/6100#discussioncomment-11549632, then this is not a "workaround" and it is by design (_this use case is one of the main reason why each hook event has its own `App` field_). Wrapping the entire request execution in a transaction by default is rarely a good idea, especially with SQLite since we allow only 1 write and/or transaction operation at a time.

In any case, my recommendation from above remains - if your use case requires more advanced or low-level SQLite primitives, then just use SQLite directly.

**Reply by patronovski** [reply]
Thanks for the explanation.
